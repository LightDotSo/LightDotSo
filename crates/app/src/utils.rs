// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::path::PathBuf;

use crate::constants::{DEV_APP_DIR, LIGHTDOTSO_APP_DIR};

// From: https://github.com/tomheaton/tauri-rspc-prisma/blob/e135a252a7c08d4a81847934ed73296c998f2753/core/src/utils.rs#L8
// License: MIT
pub fn get_lightdotso_dir() -> PathBuf {
    let path = platform_dirs::AppDirs::new(Some(&LIGHTDOTSO_APP_DIR.to_string()), true).unwrap();
    let mut data_dir = path.data_dir;

    // check if in dev mode
    if cfg!(debug_assertions) {
        data_dir.push(DEV_APP_DIR.to_string());
    }

    data_dir
}
