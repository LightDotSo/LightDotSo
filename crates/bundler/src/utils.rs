// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use ethers::types::{Address, U256};
use silius_uopool::Overhead;
/// Entire file is copied from https://github.com/Vid201/silius/blob/bc8b7b0039c9a2b02256fefc7eed3b2efc94bf96/bin/silius/src/utils.rs
/// License: MIT or Apache-2.0
use std::str::FromStr;

/// Parses address from string
pub fn parse_address(s: &str) -> Result<Address, String> {
    Address::from_str(s).map_err(|_| format!("String {s} is not a valid address"))
}

/// Parses U256 from string
pub fn parse_u256(s: &str) -> Result<U256, String> {
    U256::from_str_radix(s, 10).map_err(|_| format!("String {s} is not a valid U256"))
}

/// Helper function to calculate the call gas limit of a [UserOperation](UserOperation)
/// The function is invoked by the
/// [estimate_user_operation_gas](crates::uopool::estimate::estimate_user_operation_gas) method.
///
/// # Arguments
/// `paid` - The paid gas
/// `pre_op_gas` - The pre-operation gas
/// `fee_per_gas` - The fee per gas
///
/// # Returns
/// The call gas limit of the [UserOperation](UserOperation)
pub fn calculate_call_gas_limit(paid: U256, pre_op_gas: U256, fee_per_gas: U256) -> U256 {
    paid / fee_per_gas - pre_op_gas + Overhead::default().fixed
}
