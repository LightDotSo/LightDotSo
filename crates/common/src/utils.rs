// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use alloy::hex;
use eyre::{eyre, Result};

pub fn hex_to_bytes(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
    // Check if the hex string has the 0x prefix
    if !hex_str.starts_with("0x") {
        return Err(hex::FromHexError::InvalidStringLength);
    }

    // Remove the 0x prefix
    let pure_hex = &hex_str[2..];
    hex::decode(pure_hex) // Decode hex string to Vec<u8>
}

pub fn hex_to_bytes32(hex_str: &str) -> Result<[u8; 32]> {
    if !hex_str.starts_with("0x") {
        return Err(eyre!("Invalid hex string: no 0x prefix"));
    }

    let vec = hex_to_bytes(hex_str)?;

    if vec.len() != 32 {
        return Err(eyre!("Length mismatch: Vec is not 32 bytes long"));
    }

    let mut array = [0u8; 32];
    array.copy_from_slice(&vec);
    Ok(array)
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_hex() {
        let hex_str = "0x4a5b";
        let bytes = hex_to_bytes(hex_str).unwrap();
        assert_eq!(bytes, vec![74, 91]);
    }

    #[test]
    fn test_invalid_hex() {
        let hex_str = "0x4g";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }

    #[test]
    fn test_no_0x_prefix() {
        let hex_str = "4a5b";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }
}
