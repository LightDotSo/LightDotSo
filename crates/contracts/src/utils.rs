// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Rundler is free software: you can redistribute it and/or modify it under the
// terms of the GNU Lesser General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// Rundler is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with Rundler.
// If not, see https://www.gnu.org/licenses/.

#![allow(clippy::expect_used)]
#![allow(unused_doc_comments)]

use crate::entrypoint::EntryPoint::{ExecutionResult, FailedOp};
use alloy::{
    primitives::Bytes,
    providers::Provider,
    sol,
    sol_types::{ContractError, SolError},
};

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

/// Gets the revert data from a contract error if it is a revert error,
/// otherwise returns the original error.
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/utils/src/eth.rs#L31-37
/// License: GNU Lesser General Public License v3.0
pub fn get_revert_bytes<M: Provider>(error: ContractError<M>) -> Result<Bytes, ContractError<M>> {
    if let ContractError::Revert(revert) = error {
        Ok(Bytes::from(revert.reason().as_bytes().to_vec()))
    } else {
        Err(error)
    }
}

/// The abi for what happens when you just `revert("message")` in a contract
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/utils/src/eth.rs#L39-45
/// License: GNU Lesser General Public License v3.0
sol! {
  error Error(string);
}

/// Decode the revert data into an `ExecutionResult` or a `FailedOp`.
/// From: https://github.com/alchemyplatform/rundler/blob/ae615d0faa97b61a7e0a3d0a21793f383560ae35/crates/provider/src/ethers/entry_point.rs#L141-155
/// License: GNU Lesser General Public License v3.0
pub fn decode_simulate_handle_ops_revert(revert_data: Bytes) -> Result<ExecutionResult, String> {
    if let Ok(result) = ExecutionResult::abi_decode(&revert_data, true) {
        Ok(result)
    } else if let Ok(failed_op) = FailedOp::abi_decode(&revert_data, true) {
        Err(failed_op.reason)
    } else if let Ok(err) = Error::abi_decode(&revert_data, true) {
        Err(err._0)
    } else {
        Err(String::new())
    }
}
