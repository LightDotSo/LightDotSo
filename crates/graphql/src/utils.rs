// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::constants::THE_GRAPH_HOSTED_SERVICE_URLS;
use eyre::{eyre, Result};

/// A helper function to get the graphql url for a given chain id.
pub fn get_graphql_url(chain_id: u64) -> Result<String> {
    let url = THE_GRAPH_HOSTED_SERVICE_URLS
        .get(&chain_id)
        .ok_or_else(|| eyre!("Chain id {} not supported", chain_id))?;
    Ok(url.to_string())
}
