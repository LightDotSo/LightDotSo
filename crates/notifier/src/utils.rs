// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::types::{NotificationOperation, UserOnlyNotification, WalletOnlyNotification};
use lightdotso_prisma::{ActivityEntity, ActivityOperation};
use serde_json::Value;

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

// Utility function to match the activity entity/operation combination
pub fn match_user_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<UserOnlyNotification> {
    match (entity, operation) {
        (ActivityEntity::InviteCode, ActivityOperation::Update) => {
            // Check if the key of log is status is `USED`
            if log["status"].as_str() == Some("USED") {
                Some(UserOnlyNotification::InviteCodeAccepted)
            } else {
                None
            }
        }
        _ => None,
    }
}

// Utility function to match the activity entity/operation combination
pub fn match_wallet_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<WalletOnlyNotification> {
    match (entity, operation) {
        (ActivityEntity::UserOperation, ActivityOperation::Create) => {
            Some(WalletOnlyNotification::UserOperationCreated)
        }
        (ActivityEntity::UserOperation, ActivityOperation::Update) => {
            if log["status"].as_str() == Some("EXECUTED") ||
                log["status"].as_str() == Some("REVERTED")
            {
                Some(WalletOnlyNotification::TransactionWithUserOperationExecuted)
            } else {
                None
            }
        }
        (ActivityEntity::Transaction, ActivityOperation::Update) => None,
        _ => None,
    }
}

// Combination utility function to match the activity entity/operation combination
pub fn match_notification_operation_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<NotificationOperation> {
    if let Some(res) = match_user_only_notification_with_activity(entity, operation, log) {
        return Some(NotificationOperation::UserOnly(res));
    }

    match_wallet_only_notification_with_activity(entity, operation, log)
        .map(NotificationOperation::WalletOnly)
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    #[test]
    fn test_match_user_only_notification_with_activity() {
        let entity = ActivityEntity::InviteCode;
        let operation = ActivityOperation::Update;
        let log = serde_json::Value::from_str(r#"{"status": "USED"}"#).unwrap();
        let result = match_user_only_notification_with_activity(&entity, &operation, &log);
        assert_eq!(result, Some(UserOnlyNotification::InviteCodeAccepted));
    }

    #[test]
    fn test_match_wallet_only_notification_with_activity() {
        let entity = ActivityEntity::UserOperation;
        let operation = ActivityOperation::Create;
        let result =
            match_wallet_only_notification_with_activity(&entity, &operation, &Value::Null);
        assert_eq!(result, Some(WalletOnlyNotification::UserOperationCreated));
    }

    #[test]
    fn test_match_notification_operation_with_activity() {
        let entity = ActivityEntity::UserOperation;
        let operation = ActivityOperation::Create;
        let result = match_notification_operation_with_activity(&entity, &operation, &Value::Null);
        assert_eq!(
            result,
            Some(NotificationOperation::WalletOnly(WalletOnlyNotification::UserOperationCreated))
        );
    }
}
