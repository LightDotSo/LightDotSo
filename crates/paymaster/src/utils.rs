// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use alloy::primitives::{Address, Bytes, U256};
use eyre::Result;
use lightdotso_contracts::types::{
    EstimateResult, GasAndPaymasterAndData, PackedEstimateResult, PackedGasAndPaymasterAndData,
    PackedUserOperation, PackedUserOperationRequest, UserOperation, UserOperationRequest,
};
use lightdotso_gas::types::GasEstimation;
use lightdotso_jsonrpsee::{
    handle_response,
    types::{Request, Response},
};
use lightdotso_tracing::tracing::{info, warn};
use serde_json::{json, Value};

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

/// Get the gas and paymaster and data from the pm.
pub async fn get_gas_and_paymaster_and_data(
    rpc_url: String,
    entry_point: Address,
    user_operation: &UserOperationRequest,
    sponsorship_policy: Option<Value>,
) -> Result<Response<GasAndPaymasterAndData>> {
    let params = if let Some(policy) = sponsorship_policy {
        vec![json!(user_operation), json!(entry_point), policy]
    } else {
        vec![json!(user_operation), json!(entry_point)]
    };
    info!("params: {:?}", params);

    let req_body = Request {
        jsonrpc: "2.0".to_string(),
        method: "pm_sponsorUserOperation".to_string(),
        params: params.clone(),
        id: 1,
    };

    let client = reqwest::Client::new();
    let response = client.post(rpc_url).json(&req_body).send().await?;

    // Handle the response for the JSON-RPC API.
    handle_response(response).await
}

/// Get the packed gas and paymaster and data from the pm.
pub async fn get_packed_gas_and_paymaster_and_data(
    rpc_url: String,
    entry_point: Address,
    packed_user_operation: &PackedUserOperationRequest,
) -> Result<Response<PackedGasAndPaymasterAndData>> {
    let params = vec![json!(packed_user_operation), json!(entry_point)];
    info!("params: {:?}", params);

    let req_body = Request {
        jsonrpc: "2.0".to_string(),
        method: "pm_sponsorUserOperation".to_string(),
        params: params.clone(),
        id: 1,
    };

    let client = reqwest::Client::new();
    let response = client.post(rpc_url).json(&req_body).send().await?;

    // Handle the response for the JSON-RPC API.
    handle_response(response).await
}

/// Estimate the gas for the request w/ the internal gas API.
pub async fn estimate_request_gas_estimation(chain_id: u64) -> Result<Response<GasEstimation>> {
    let params = vec![chain_id];
    info!("params: {:?}", params);

    let req_body = Request {
        jsonrpc: "2.0".to_string(),
        method: "gas_requestGasEstimation".to_string(),
        params: params.clone(),
        id: 1,
    };

    let client = reqwest::Client::new();
    let response =
        client.post("http://lightdotso-gas.internal:3000").json(&req_body).send().await?;

    // Handle the response for the JSON-RPC API.
    handle_response(response).await
}

/// From: https://github.com/qi-protocol/ethers-userop/blob/50cb1b18a551a681786f1a766d11215c80afa7cf/src/userop_middleware.rs#L128
/// License: MIT
pub async fn estimate_user_operation_gas(
    chain_id: u64,
    entry_point: Address,
    user_operation: &UserOperationRequest,
) -> Result<Response<EstimateResult>> {
    let params = vec![json!(user_operation), json!(entry_point)];
    info!("params: {:?}", params);

    let req_body = Request {
        jsonrpc: "2.0".to_string(),
        method: "eth_estimateUserOperationGas".to_string(),
        params: params.clone(),
        id: 1,
    };

    let client = reqwest::Client::new();
    let response = client
        .post(format!("http://lightdotso-rpc-internal.internal:3000/internal/{}", chain_id))
        .json(&req_body)
        .send()
        .await?;

    // Handle the response for the JSON-RPC API.
    handle_response(response).await
}

/// Estimate the gas for the packed user operation.
pub async fn estimate_packed_user_operation_gas(
    chain_id: u64,
    entry_point: Address,
    packed_user_operation: PackedUserOperationRequest,
) -> Result<Response<PackedEstimateResult>> {
    let params = vec![json!(packed_user_operation), json!(entry_point)];
    info!("params: {:?}", params);

    let req_body = Request {
        jsonrpc: "2.0".to_string(),
        method: "eth_estimateUserOperationGas".to_string(),
        params: params.clone(),
        id: 1,
    };

    let client = reqwest::Client::new();
    let response = client
        .post(format!("http://lightdotso-rpc-internal.internal:3000/internal/{}", chain_id))
        .json(&req_body)
        .send()
        .await?;

    // Handle the response for the JSON-RPC API.
    handle_response(response).await
}

/// Construct the user operation w/ rpc.
pub async fn construct_user_operation(
    chain_id: u64,
    user_operation: UserOperationRequest,
    entry_point: Address,
) -> Result<UserOperation> {
    // If the `preVerificationGas`, `verificationGasLimit`, and `callGasLimit` are set,
    // override the gas estimation for the user operation
    let estimated_user_operation_gas: EstimateResult = if user_operation
        .pre_verification_gas
        .is_some_and(|pre_verification_gas| pre_verification_gas > U256::ZERO) &&
        user_operation
            .verification_gas_limit
            .is_some_and(|verification_gas_limit| verification_gas_limit > U256::ZERO) &&
        user_operation.call_gas_limit.is_some_and(|call_gas_limit| call_gas_limit > U256::ZERO)
    {
        warn!("Overriding the gas estimation for the user operation");
        EstimateResult {
            pre_verification_gas: user_operation.pre_verification_gas.unwrap_or_default(),
            verification_gas_limit: user_operation.verification_gas_limit.unwrap_or_default(),
            call_gas_limit: user_operation.call_gas_limit.unwrap_or_default(),
        }
    } else {
        // If the `estimate_user_operation_gas` is not set, estimate the gas for the user operation.
        estimate_user_operation_gas(chain_id, entry_point, &user_operation).await?.result
    };
    info!("estimated_user_operation_gas: {:?}", estimated_user_operation_gas);

    // If the `maxFeePerGas` and `maxPriorityFeePerGas` are set, include them in the user operation.
    if user_operation.max_fee_per_gas.is_some_and(|max_fee_per_gas| max_fee_per_gas > U256::ZERO) &&
        user_operation
            .max_priority_fee_per_gas
            .is_some_and(|max_priority_fee_per_gas| max_priority_fee_per_gas > U256::ZERO)
    {
        warn!("Overriding the gas estimation for the user operation w/ the maxFeePerGas and maxPriorityFeePerGas");
        return Ok(UserOperation {
            call_data: user_operation.call_data,
            init_code: user_operation.init_code,
            nonce: user_operation.nonce,
            sender: user_operation.sender,
            call_gas_limit: estimated_user_operation_gas.call_gas_limit,
            verification_gas_limit: estimated_user_operation_gas.verification_gas_limit,
            pre_verification_gas: estimated_user_operation_gas.pre_verification_gas,
            max_fee_per_gas: user_operation.max_fee_per_gas.unwrap_or_default(),
            max_priority_fee_per_gas: user_operation.max_priority_fee_per_gas.unwrap_or_default(),
            signature: user_operation.signature,
            paymaster_and_data: Bytes::default(),
        });
    }

    // Get the estimated request gas because required gas parameters are not set.
    let estimated_request_gas = estimate_request_gas_estimation(chain_id).await?.result;

    Ok(UserOperation {
        call_data: user_operation.call_data,
        init_code: user_operation.init_code,
        nonce: user_operation.nonce,
        sender: user_operation.sender,
        call_gas_limit: estimated_user_operation_gas.call_gas_limit,
        verification_gas_limit: estimated_user_operation_gas.verification_gas_limit,
        pre_verification_gas: estimated_user_operation_gas.pre_verification_gas,
        max_fee_per_gas: estimated_request_gas.high.max_fee_per_gas,
        max_priority_fee_per_gas: estimated_request_gas.high.max_priority_fee_per_gas,
        signature: user_operation.signature,
        paymaster_and_data: Bytes::default(),
    })
}

/// Construct the packed user operation.
pub async fn construct_packed_user_operation(
    chain_id: u64,
    packed_user_operation: PackedUserOperationRequest,
    entry_point: Address,
) -> Result<PackedUserOperation> {
    // If the `preVerificationGas`, `verificationGasLimit`, and `callGasLimit` are set,
    // override the gas estimation for the user operation
    let estimated_packed_user_operation_gas: PackedEstimateResult = if packed_user_operation
        .pre_verification_gas
        .is_some_and(|pre_verification_gas| pre_verification_gas > U256::ZERO) &&
        packed_user_operation
            .verification_gas_limit
            .is_some_and(|verification_gas_limit| verification_gas_limit > U256::ZERO) &&
        packed_user_operation
            .call_gas_limit
            .is_some_and(|call_gas_limit| call_gas_limit > U256::ZERO)
    {
        warn!("Overriding the gas estimation for the user operation");
        PackedEstimateResult {
            pre_verification_gas: packed_user_operation
                .clone()
                .pre_verification_gas
                .unwrap_or_default(),
            verification_gas_limit: packed_user_operation
                .clone()
                .verification_gas_limit
                .unwrap_or_default(),
            call_gas_limit: packed_user_operation.clone().call_gas_limit.unwrap_or_default(),
            paymaster_verification_gas_limit: packed_user_operation
                .clone()
                .paymaster_verification_gas_limit
                .unwrap_or_default(),
        }
    } else {
        // If the `estimate_packed_user_operation_gas` is not set, estimate the gas for the packed
        // user operation.
        estimate_packed_user_operation_gas(chain_id, entry_point, packed_user_operation.clone())
            .await?
            .result
    };
    info!("estimated_packed_user_operation_gas: {:?}", estimated_packed_user_operation_gas);

    // If the `maxFeePerGas` and `maxPriorityFeePerGas` are set, include them in the user operation.
    if packed_user_operation
        .clone()
        .max_fee_per_gas
        .is_some_and(|max_fee_per_gas| max_fee_per_gas > U256::ZERO) &&
        packed_user_operation
            .clone()
            .max_priority_fee_per_gas
            .is_some_and(|max_priority_fee_per_gas| max_priority_fee_per_gas > U256::ZERO)
    {
        warn!("Overriding the gas estimation for the user operation w/ the maxFeePerGas and maxPriorityFeePerGas");
        return Ok(PackedUserOperation {
            call_data: packed_user_operation.clone().call_data,
            factory: packed_user_operation.clone().factory,
            factory_data: packed_user_operation.clone().factory_data,
            nonce: packed_user_operation.clone().nonce,
            sender: packed_user_operation.clone().sender,
            call_gas_limit: estimated_packed_user_operation_gas.call_gas_limit,
            verification_gas_limit: estimated_packed_user_operation_gas.verification_gas_limit,
            pre_verification_gas: estimated_packed_user_operation_gas.pre_verification_gas,
            max_fee_per_gas: packed_user_operation.clone().max_fee_per_gas.unwrap_or_default(),
            max_priority_fee_per_gas: packed_user_operation
                .clone()
                .max_priority_fee_per_gas
                .unwrap_or_default(),
            signature: packed_user_operation.clone().signature,
            paymaster: packed_user_operation.clone().paymaster,
            paymaster_verification_gas_limit: Some(
                packed_user_operation.clone().paymaster_verification_gas_limit.unwrap_or_default(),
            ),
            paymaster_post_op_gas_limit: Some(
                packed_user_operation.clone().paymaster_post_op_gas_limit.unwrap_or_default(),
            ),
            paymaster_data: packed_user_operation.clone().paymaster_data,
        });
    }

    // Get the estimated request gas because required gas parameters are not set.
    let estimated_request_gas = estimate_request_gas_estimation(chain_id).await?.result;

    Ok(PackedUserOperation {
        call_data: packed_user_operation.clone().call_data,
        factory: packed_user_operation.clone().factory,
        factory_data: packed_user_operation.clone().factory_data,
        nonce: packed_user_operation.clone().nonce,
        sender: packed_user_operation.clone().sender,
        call_gas_limit: estimated_packed_user_operation_gas.call_gas_limit,
        verification_gas_limit: estimated_packed_user_operation_gas.verification_gas_limit,
        pre_verification_gas: estimated_packed_user_operation_gas.pre_verification_gas,
        max_fee_per_gas: estimated_request_gas.high.max_fee_per_gas,
        max_priority_fee_per_gas: estimated_request_gas.high.max_priority_fee_per_gas,
        signature: packed_user_operation.clone().signature,
        paymaster: packed_user_operation.clone().paymaster,
        paymaster_verification_gas_limit: Some(
            packed_user_operation.clone().paymaster_verification_gas_limit.unwrap_or_default(),
        ),
        paymaster_post_op_gas_limit: Some(
            packed_user_operation.clone().paymaster_post_op_gas_limit.unwrap_or_default(),
        ),
        paymaster_data: packed_user_operation.clone().paymaster_data,
    })
}
