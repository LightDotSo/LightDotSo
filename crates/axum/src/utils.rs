// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use ethers_main::utils::hex;
use eyre::{eyre, Result};

pub fn hex_to_bytes(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
    // Check if the hex string has the 0x prefix
    if !hex_str.starts_with("0x") {
        return Err(hex::FromHexError::InvalidStringLength);
    }

    // Remove the 0x prefix
    let pure_hex = &hex_str[2..];
    hex::decode(pure_hex) // Decode hex string to Vec<u8>
}

pub fn hex_to_bytes32(hex_str: &str) -> Result<[u8; 32]> {
    if !hex_str.starts_with("0x") {
        return Err(eyre!("Invalid hex string: no 0x prefix"));
    }

    let vec = hex_to_bytes(hex_str)?;

    if vec.len() != 32 {
        return Err(eyre!("Length mismatch: Vec is not 32 bytes long"));
    }

    let mut array = [0u8; 32];
    array.copy_from_slice(&vec);
    Ok(array)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_hex() {
        let hex_str = "0x4a5b";
        let bytes = hex_to_bytes(hex_str).unwrap();
        assert_eq!(bytes, vec![74, 91]);
    }

    #[test]
    fn test_invalid_hex() {
        let hex_str = "0x4g";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }

    #[test]
    fn test_no_0x_prefix() {
        let hex_str = "4a5b";
        let result = hex_to_bytes(hex_str);
        assert!(result.is_err());
    }
}
