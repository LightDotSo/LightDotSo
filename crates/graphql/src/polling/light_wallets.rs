// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::utils::get_graphql_url;
use eyre::Result;

#[cynic::schema("graph")]
mod schema {}

#[derive(cynic::QueryVariables, Debug)]
pub struct GetLightWalletsQueryVariables {
    pub min_block: BigInt,
    pub min_index: BigInt,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "GetLightWalletsQueryVariables")]
pub struct GetLightWalletsQuery {
    #[arguments(where: { blockNumber_gte: $min_block, index_gt: $min_index }, orderBy: "index")]
    pub light_wallets: Vec<LightWallet>,
    pub _meta: Option<Meta>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "_Meta_")]
pub struct Meta {
    pub block: Block,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "_Block_")]
pub struct Block {
    pub number: i32,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct LightWallet {
    pub id: Bytes,
    pub index: BigInt,
    pub address: Bytes,
    pub image_hash: Option<Bytes>,
    pub user_op_hash: Bytes,
    pub sender: Bytes,
    pub factory: Bytes,
    pub paymaster: Bytes,
    pub block_number: BigInt,
    pub block_timestamp: BigInt,
    pub transaction_hash: Bytes,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[cynic(graphql_type = "LightWallet_orderBy")]
pub enum LightWalletOrderBy {
    #[cynic(rename = "id")]
    Id,
    #[cynic(rename = "index")]
    Index,
    #[cynic(rename = "address")]
    Address,
    #[cynic(rename = "imageHash")]
    ImageHash,
    #[cynic(rename = "userOpHash")]
    UserOpHash,
    #[cynic(rename = "sender")]
    Sender,
    #[cynic(rename = "factory")]
    Factory,
    #[cynic(rename = "paymaster")]
    Paymaster,
    #[cynic(rename = "blockNumber")]
    BlockNumber,
    #[cynic(rename = "blockTimestamp")]
    BlockTimestamp,
    #[cynic(rename = "transactionHash")]
    TransactionHash,
}

#[derive(cynic::Scalar, Debug, Clone)]
pub struct BigInt(pub String);

#[derive(cynic::Scalar, Debug, Clone)]
pub struct Bytes(pub String);
pub fn build_light_wallets_query(
    vars: GetLightWalletsQueryVariables,
) -> cynic::Operation<GetLightWalletsQuery, GetLightWalletsQueryVariables> {
    use cynic::QueryBuilder;

    GetLightWalletsQuery::build(vars)
}

pub fn run_light_wallets_query(
    chain_id: u64,
    vars: GetLightWalletsQueryVariables,
) -> Result<cynic::GraphQlResponse<GetLightWalletsQuery>> {
    use cynic::http::ReqwestBlockingExt;

    let query = build_light_wallets_query(vars);

    Ok(reqwest::blocking::Client::new().post(get_graphql_url(chain_id)?).run_graphql(query)?)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn snapshot_test_query() {
        // Running a snapshot test of the query building functionality as that gives us
        // a place to copy and paste the actual GQL we're using for running elsewhere,
        // and also helps ensure we don't change queries by mistake

        let query = build_light_wallets_query(GetLightWalletsQueryVariables {
            min_block: BigInt("0".to_string()),
            min_index: BigInt("0".to_string()),
        });

        insta::assert_snapshot!(query.query);
    }

    #[test]
    fn test_running_query() {
        let result = run_light_wallets_query(
            1,
            GetLightWalletsQueryVariables {
                min_block: BigInt("0".to_string()),
                min_index: BigInt("0".to_string()),
            },
        )
        .unwrap();
        if result.errors.is_some() {
            assert_eq!(result.errors.unwrap().len(), 0);
        }
    }
}
