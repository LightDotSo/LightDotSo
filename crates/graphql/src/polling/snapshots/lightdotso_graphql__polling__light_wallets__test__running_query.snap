---
source: crates/graphql/src/polling/light_wallets.rs
expression: result.data
---
Some(
    GetProtocolQuery {
        light_wallets: [],
        _meta: Some(
            Meta {
                block: Block {
                    number: 18253429,
                },
            },
        ),
    },
)
